skfda/misc/operators/_linear_differential_operator.py

```diff
@@ -585,42 +635,76 @@ def fdatagrid_penalty_matrix_optimized(
     """
     Optimized version for FDatagrid.
 
-    If the data_matrix of the FDataGrid is the identity, the resulting
-    matrix will be the gram matrix for functions discretized in the
-    given grid points. That is to say, in order to calculate the norm
+    The datamatrix of the given FDataGrid has to be the identity matrix,
+    the resulting matrix will be the gram matrix for functions discretized
+    in the given grid points. That is to say, in order to calculate the norm
     of the linear operator applied to functions discretized in the given
     grid points, it is enough to multiply the returned matrix by the
     values in the points of the grid on both sides.
 
-    If the data_matrix of the FDataGrid is not the identity, the resulting
-    matrix will be the gram matrix for functions expressed as a combination
-    of the given set of functions (the entries of the fdatagrid). The
-    intended use is to calculate the norm of the linear operator for
-    functions expressed in a basis that is a FDataGrid (see CustomBasis).
-
-    Note that the first case is a particular case of the second one.
     """
-    evaluated_basis = sum(
-        w(basis.grid_points[0]) if callable(w) else w
-        * basis.derivative(order=i)(basis.grid_points[0])
-        for i, w in enumerate(linear_operator.weights)
-    )
-    assert not isinstance(evaluated_basis, int)
-
-    indices = np.triu_indices(basis.n_samples)
-    product = evaluated_basis[indices[0]] * evaluated_basis[indices[1]]
+    max_deriv = len(linear_operator.weights) - 1
+    n_points = basis.grid_points[0].shape[0]
+    findiff_accuracy = 2
 
-    triang_vec = scipy.integrate.simps(product[..., 0], x=basis.grid_points)
-
-    matrix = np.empty((basis.n_samples, basis.n_samples))
-
-    # Set upper matrix
-    matrix[indices] = triang_vec
+    evaluated_basis = _optimized_operator_evaluation_in_grid(
+        linear_operator,
+        basis.grid_points[0],
+        findiff_accuracy=findiff_accuracy,
+    )
 
-    # Set lower matrix
-    matrix[(indices[1], indices[0])] = triang_vec
+    # The support of the result of applying the linear operator
+    # is a small subset of the grid points. We calculate the
+    # maximum number of points to the left and right of the
+    # point where the linear operator is applied that can be
+    # different from zero.
+
+    # Formula derived from the definition of coefficients_not_uni
+    # in findiff.
+    spread = 2 * math.floor((max_deriv + 1) / 2) - 1
+    # The accuracy parameter in findiff increases the number of
+    # points considered to the left and the right
+    spread += findiff_accuracy + 1
+
+    product_matrix = np.zeros((basis.n_samples, basis.n_samples))
+    for i in range(n_points):
```

Comment:
This logic is very complicated. I also feel that is repeating nontrivial work that was already done in `_optimized_operator_evaluation_in_grid`.

Reply:
The new logic should be less complicated. It should be readable now. I think the complexity might have moved into `_optimized_operator_evaluation_in_grid`. I tried to mitigate it with some comments, as the meaning of the coefficients is not as straightforward as I would like. 
The issue is that the coefficients for the $i$-th point are intended to give you a way to calculate the value of the operator at that point. This is a similar problem to calculating the result of applying the operator to the delta basis. However, it is not entirely immediate (at least for me) that concatenating the coefficients column-wise and reading $j$-th row of the resulting matrix leads to the valuex of applying the operator to the $j$-th function of the delta basis.